kong/api/routes/upstreams.lua

```diff
@@ -133,40 +175,48 @@ return {
     GET = function(self, dao_factory)
       self.params.active = nil
 
-      local target_history, err = dao_factory.targets:find_all({
-        upstream_id = self.params.upstream_id,
-      })
-      if not target_history then
-        return app_helpers.yield_error(err)
-      end
+      local active, active_n = get_active_targets(dao_factory,
+                                                  self.params.upstream_id)
 
-      --sort and walk based on target and creation time
-      for _, target in ipairs(target_history) do
-        target.order = target.target .. ":" ..
-          target.created_at .. ":" .. target.id
-      end
-      table.sort(target_history, function(a, b) return a.order > b.order end)
+      -- for now lets not worry about rolling our own pagination
+      -- we also end up returning a "backwards" list of targets because
+      -- of how we sorted- do we care?
+      return responses.send_HTTP_OK {
+        total = active_n,
+        data  = active,
+      }
+    end,
 
-      local seen     = {}
-      local active   = setmetatable({}, cjson.empty_array_mt)
-      local active_n = 0
+    POST = function(self, dao_factory, helpers)
+      clean_history(self.params.upstream_id, dao_factory)
 
-      for _, entry in ipairs(target_history) do
-        if not seen[entry.target] then
-          if entry.weight == 0 then
-            seen[entry.target] = true
+      crud.post(self.params, dao_factory.targets)
+    end,
+  },
 
-          else
-            entry.order = nil -- dont show our order key to the client
+  ["/upstreams/:upstream_name_or_id/health/"] = {
+    before = function(self, dao_factory, helpers)
+      crud.find_upstream_by_name_or_id(self, dao_factory, helpers)
+      self.params.upstream_id = self.upstream.id
+    end,
 
-            -- add what we want to send to the client in our array
-            active_n = active_n + 1
-            active[active_n] = entry
+    GET = function(self, dao_factory)
+      self.params.active = nil
+
+      local upstream_id = self.params.upstream_id
+      local active, active_n = get_active_targets(dao_factory, upstream_id)
+
+      local health_info, err = balancer.get_upstream_health(upstream_id)
+      if err then
+        ngx.log(ngx.ERR, "[healthchecks] failed getting upstream health: ", err)
+      end
 
-            -- track that we found this host:port so we only show
-            -- the most recent one (kinda)
-            seen[entry.target] = true
-          end
+      for _, entry in ipairs(active) do
+        if health_info then
+          entry.health = health_info[entry.target] or "not in balancer"
+        else
+          -- Healthchecks disabled for this upstream
+          entry.health = ngx.null
```

Comment:
Should we here distinguish cases where the above `get_upstream_health` returned an error, from cases where the healthchecks were disabled? I feel like logging an error in Kong might not be enough, the user probably should be informed that "something went wrong" while retrieving the health of those particular targets.